### PR TITLE
[fix]: `dynamic_programming/cut_rod.cpp` does not compile

### PR DESCRIPTION
**Pull request by Pardeep009**
#### Description of Change

included climits library in code to remove compilation error.
#### Checklist

- [x] Added description of change
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1085"></a>

### CODE REVIEW

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1085/checks?check_run_id=1154679221) warnings.

**Comment by Pardeep009** [comment]
@Panquesito7 can you please specify me, which changes do you want in this code?
any example?

File: dynamic_programming/cut_rod.cpp [hunk 1]
```diff
@@ -3,23 +3,33 @@ contains prices of all pieces of size smaller than n. Determine
 the maximum value obtainable by cutting up the rod and selling
 the pieces.*/
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:4]: See #962 as a reference for how to document the first comment block properly.

File: dynamic_programming/cut_rod.cpp [hunk 2]
```diff
@@ -3,23 +3,33 @@ contains prices of all pieces of size smaller than n. Determine
...
 
+#include <array>
+#include <climits>
 #include <iostream>
-using namespace std;
-int cutrod(int p[], int n) {
-    int r[n + 1];
+
+using std::cout;
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:10]: Better to use the namespace directly.
```suggestion
```

File: dynamic_programming/cut_rod.cpp [hunk 3]
```diff
@@ -3,23 +3,33 @@ contains prices of all pieces of size smaller than n. Determine
...
 }
 int main() {
-    int price[] = {1,  5,  8,  9,  10, 17, 17, 20, 24, 30, 31, 32, 33, 34, 35,
-                   36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50};
-    cout << cutrod(price, 30);
+    const int n = 30;
+    std::array<int, n> price = {1,  5,  8,  9,  10, 17, 17, 20, 24, 30,
+                                31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
+                                41, 42, 43, 44, 45, 46, 47, 48, 49, 50};
+    cout << cutrod(price, n);
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:33]:
```suggestion
    std::cout << cutrod(price, n);
```

File: dynamic_programming/cut_rod.cpp [hunk 4]
```diff
@@ -3,23 +3,33 @@ contains prices of all pieces of size smaller than n. Determine
...
+#include <climits>
 #include <iostream>
-using namespace std;
-int cutrod(int p[], int n) {
-    int r[n + 1];
+
+using std::cout;
+
+template <size_t T>
+
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:13]:
```suggestion
```

File: dynamic_programming/cut_rod.cpp [hunk 5]
```diff
@@ -3,23 +3,33 @@ contains prices of all pieces of size smaller than n. Determine
...
+int cutrod(const std::array<int, T> &p, const int n) {
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:14]: Function not documented.
See #962 as a reference.

File: dynamic_programming/cut_rod.cpp [hunk 6]
```diff
@@ -3,23 +3,33 @@ contains prices of all pieces of size smaller than n. Determine
...
+            q = std::max(q, p[i] + r[j - i]);
         }
         r[j + 1] = q;
     }
-    return r[n];
+    int ans = r[n];
+    delete[] r;
+    return ans;
 }
 int main() {
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:28]:
```suggestion

/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by Pardeep009** [comment]
Thank you so much for this brief explanation. I will correct them in the next commit.

**Review by Panquesito7 — COMMENTED** [review]
Why delete `dynamic_programming/kadane2.cpp`?

File: dynamic_programming/kadane2.cpp [hunk 7]
```diff
@@ -70,4 +71,4 @@ int main() {
     std::cout << "Maximum subarray sum is " << max_sum;  // Printing the answer
 
     return 0;
-}
+}
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:74]:
```suggestion
}

```

File: dynamic_programming/kadane2.cpp [hunk 8]
```diff
@@ -27,7 +27,8 @@
 namespace dynamic_programming {
 /**
  * @namespace kadane
- * @brief Functions for [Kadane](https://en.wikipedia.org/wiki/Kadane%27s_algorithm) algorithm.
+ * @brief Functions for
+ * [Kadane](https://en.wikipedia.org/wiki/Kadane%27s_algorithm) algorithm.
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:31]:
```suggestion
 * @brief Functions for [Kadane](https://en.wikipedia.org/wiki/Kadane%27s_algorithm) algorithm.
```

**Comment by Pardeep009** [comment]
> Why delete `dynamic_programming/kadane2.cpp`?

By mistake, I have deleted it. Now it is added back.

File: dynamic_programming/cut_rod.cpp [hunk 9]
```diff
@@ -1,25 +1,103 @@
-/*Given a rod of length n inches and an array of prices that
-contains prices of all pieces of size smaller than n. Determine
-the maximum value obtainable by cutting up the rod and selling
-the pieces.*/
+/**
+ * @file cut_rod.cpp
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:2]:
```suggestion
 * @file
```

File: dynamic_programming/cut_rod.cpp [hunk 10]
```diff
@@ -1,25 +1,103 @@
...
-    return r[n];
+    int ans = profit[n];
+    delete[] profit;
+    return ans;  // returning maximum profit
+}
+}  // namespace cut_rod
+}  // namespace dynamic_programming
+
+/**
+ * Function to test above algorithm
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:68]:
```suggestion
 * @brief Function to test above algorithm
 * @returns void
```

File: dynamic_programming/cut_rod.cpp [hunk 11]
```diff
@@ -1,25 +1,103 @@
...
+ */
+void test() {
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:70]:
```suggestion
static void test() {
```

File: dynamic_programming/cut_rod.cpp [hunk 12]
```diff
@@ -1,25 +1,103 @@
-/*Given a rod of length n inches and an array of prices that
-contains prices of all pieces of size smaller than n. Determine
-the maximum value obtainable by cutting up the rod and selling
-the pieces.*/
+/**
+ * @file cut_rod.cpp
+ * @brief Implementation of cutting a rod problem
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:3]: Provide a Wikipedia link if possible, in Markdown format.

File: dynamic_programming/cut_rod.cpp [hunk 13]
```diff
@@ -1,25 +1,103 @@
...
+ */
+namespace dynamic_programming {
+/**
+ * @namespace cut_rod
+ * @brief Implementation of cutting a rod problem
+ */
+namespace cut_rod {
+/**
+ * @brief cut_rod function cuts the rod in different pieces and stores the
+ * maximum profit for each piece of the rod.
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:38]: Missing template parameter documentation.
Confused? See #962 as a reference.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Excellent work, code looks much more refined now. 👍
Just some missing tweaks in the code needed.

File: dynamic_programming/cut_rod.cpp (hunk 12 above)
Comment (Pardeep009) [reply dynamic_programming/cut_rod.cpp:3]: I was also finding the relevant Wikipedia page but didn't found anything. If found will update in the future.

File: dynamic_programming/cut_rod.cpp (hunk 13 above)
Comment (Panquesito7) [reply dynamic_programming/cut_rod.cpp:38]: For example:

```c++
* @tparam N number of array size
```

Comment (Pardeep009) [reply dynamic_programming/cut_rod.cpp]: If you understand this code, can you add it on your own, I am not getting it.

File: dynamic_programming/cut_rod.cpp [hunk 14]
```diff
@@ -1,25 +1,104 @@
...
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @namespace cut_rod
+ * @brief Implementation of cutting a rod problem
+ */
+namespace cut_rod {
+/**
+ * @brief maxProfitByCuttingRod function cuts the rod in different pieces and
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:37]:
```suggestion
 * @brief Cuts the rod in different pieces and
```

File: dynamic_programming/cut_rod.cpp [hunk 15]
```diff
@@ -1,25 +1,104 @@
...
+ * stores the maximum profit for each piece of the rod.
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:38]: Please document what the template parameter `T` does.
```suggestion
 * stores the maximum profit for each piece of the rod.
 * @tparam T description
```

**Review by Panquesito7 — COMMENTED** [review]
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/Pardeep009/C-Plus-Plus/actions.

**Comment by Pardeep009** [comment]
> Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/Pardeep009/C-Plus-Plus/actions.

enabled

**Comment by Panquesito7** [comment]
Please make a small commit to trigger the GitHub Actions. For example, remove line 19.
If GitHub Actions is enabled, we'll see a commit by GitHub actions updating `DIRECTORY.md`. 🙂

**Review by Panquesito7 — APPROVED** [review]
LGTM, excellent work. 👍 😄

File: dynamic_programming/cut_rod.cpp (hunk 12 above)
Comment (kvedala) [reply dynamic_programming/cut_rod.cpp:3]: You can link to other reqources like:
From universities: 
[this](https://web.stanford.edu/class/archive/cs/cs161/cs161.1168/lecture12.pdf)
[this](https://www.radford.edu/~nokie/classes/360/dp-rod-cutting.html)

Other: [this](https://www.educative.io/edpresso/the-rod-cutting-problem)

**Review by kvedala — APPROVED** [review]
LGTM 👍

**Comment by Pardeep009** [comment]
I would like to update codes and add documentation in the remaining codes of the dynamic programming section, should iIgo for that?

**Comment by Panquesito7** [comment]
Sounds great, please make one PR per file.
Thank you for your interest in contributing! 🙂

**Comment by Pardeep009** [comment]
> Sounds great, please make one PR per file.
> Thank you for your interest in contributing!

yes sure

**Pull request merged**